### PR TITLE
Fix codegen with the latest xcb-proto and update

### PR DESCRIPTION
**Pull request by Stebalien**
I started down this path because I wanted to be able to fix the ash/lsh issue then run `make`. But I
ran into a bunch of issues running `make`. This patch set fixes those issues:

1. Implement float & double types (for the GLX extension). Not the prettiest code, but it works...
2. Improve type-name resolution.
3. Handle `<length>` elements (I think).
4. Switch back to the Emacs 28 pretty-printer format (when on Emacs 29+).
5. Run `make` with xcb-proto 1.16.0.

See the commit messages for more detail, and review commit by commit.

### CODE REVIEW

**Comment by Stebalien** [comment]
The float conversion code is going to take some work...

**Comment by minad** [comment]
Thank you for working on this. Regenerating the files was also on my todo list.

**Comment by Stebalien** [comment]
Still working on writing an actually working IEEE754 implementation in elisp, but I'm getting closer.

File: xcb-types.el [hunk 1]
```diff
@@ -456,6 +461,66 @@ Consider let-bind it rather than change its global value."))
             (throw 'break 'nil)))))
     result))
 
+
+(defsubst xcb:-f64-to-binary64 (value)
```
Comment (Stebalien) [inline xcb-types.el:465]: There's some code duplication here, and lots of constants:

1. Abstracting over f32/f64 made reading the code impossible.
2. The constants might as well be magic numbers. The calculations weren't any easier to understand.

**Comment by Stebalien** [comment]
Test cases:

```elisp
(defconst f32-test-cases '((#x00000001 . 1.4012984643e-45)
                                (#x007fffff . 1.1754942107e-38)
                                (#x00800000 . 1.17549435083e-38)
                                (#x7f7fffff . 3.4028234664e38)
                                (#x3f7fffff . 0.999999940395355225)
                                (#x3f800000 . 1.0)
                                (#x3f800001 . 1.00000011920928955)
                                (#xc0000000 . -2.0)
                                (#x00000000 . 0.0)
                                (#x80000000 . -0.0)
                                (#x7f800000 . 1e+INF)
                                (#xff800000 . -1e+INF)
                                (#x40490fdb . 3.14159274101257324)
                                (#x3eaaaaab . 0.333333343267440796)))
(let ((ctr 0))
  (pcase-dolist (`(,b . ,f) f32-test-cases)
    (let ((to (xcb:-f32-to-binary32 f))
          (from (xcb:-binary32-to-f32 b)))
      (unless (= b to) (print (format "%d: %s -> %x != %x" ctr f to b)))
      (unless (= f from) (print (format "%d: %x -> %s != %s" ctr b from f))))
    (cl-incf ctr)))

(defconst f64-test-cases `((#x3ff0000000000000 . 1.0)
                                (#x3ff0000000000001 . 1.0000000000000002)
                                (#x3ff0000000000002 . 1.0000000000000004)
                                (#x4000000000000000 . 2.0)
                                (#xc000000000000000 . -2.0)
                                (#x4008000000000000 . 3.0)
                                (#x4010000000000000 . 4.0)
                                (#x4014000000000000 . 5.0)
                                (#x4018000000000000 . 6.0)
                                (#x4037000000000000 . 23.0)
                                (#x3f88000000000000 . 0.01171875)
                                (#x0000000000000001 . 4.9406564584124654e-324)
                                (#x000fffffffffffff . 2.2250738585072009e-308)
                                (#x0010000000000000 . 2.2250738585072014e-308)
                                (#x7fefffffffffffff . 1.7976931348623157e308)
                                (#x0000000000000000 . +0.0)
                                (#x8000000000000000 . -0.0)
                                (#x7ff0000000000000 . +1e+INF)
                                (#xfff0000000000000 . -1e+INF)
                                (#x3fd5555555555555 . ,(/ 1.0 3.0))
                                (#x400921fb54442d18 . ,float-pi)))
(let ((ctr 0))
  (pcase-dolist (`(,b . ,f) f64-test-cases)
    (let ((to (xcb:-f64-to-binary64 f))
          (from (xcb:-binary64-to-f64 b)))
      (unless (= b to) (print (format "%d: %s -> %x != %x" ctr f to b)))
      (unless (= f from) (print (format "%d: %x -> %s != %s" ctr b from f))))
    (cl-incf ctr)))
```

Ish... there are some rounding issues. I'll research unit testing in Elisp later unless you happen to have a system you prefer.

File: xcb-types.el (hunk 1 above)
Comment (minad) [reply xcb-types.el:465]: Doesn't look that bad. The constants look familiar.

**Comment by minad** [comment]
> Ish... there are some rounding issues. I'll research unit testing in Elisp later unless you happen to have a system you prefer.

I prefer ERT. This is the system used by Emacs itself and also by Compat, see https://github.com/emacs-compat/compat/blob/main/compat-tests.el. See also the Github workflow and Makefile there. I am at least somewhat familiar with ERT.

**Comment by Stebalien** [comment]
SG. But I'd like to consider re-organizing this a bit before adding test files https://github.com/emacs-exwm/xelb/pull/11.

**Comment by Stebalien** [comment]
>> Implement float & double types (for the GLX extension). Not the prettiest code, but it works...

> Can't review, sorry. (If you think my review is crucial let me know and I'll block some time to do it.)

Don't even try, that's why I wrote the tests.

> I find it disconcerting that the ~size field is output twice, once with :initarg and once with :initform. Could both :initarg and :initform be placed in a single form?

Probably... I thought I did it that way mimicking other code, but I can't remember now.

I'm actually going to do what `xcb:-enum` does (kind of) and declare `size` in the base class with an initform of nil.

> I'm not sure whether this is completely correct. When is :initform evaluated? have all depended upon fields (len for DeviceClass) been initialized by then?

My code is just wrong. The slot will be bound to the initform _unevaluated_, but:

1. It still needs to be evaluated (lazily).
3. `xcb:unmarshal` for `xcb:-struct` needs to handle it.

**Comment by Stebalien** [comment]
(new code should be fixed, but I won't be able to test until tonight)

**Comment by minad** [comment]
> I mentioned somewhere that I worked on lsh and ash.

Regarding the `ash`, `lsh` changes, I also inspected the code briefly and I haven't found suspicious uses of `lsh`, but I will check again carefully.

> This is good, thank you. Switching to the new format should happen in a separate commit. Not sure when we should do that. Now?

I wouldn't mind changing to a the format. The generator script could then check if emacs-major-version is 29.

I am a bit short on time right now, but I may be able to give this PR a thorough test later and also review it in greater detail.

**Comment by Stebalien** [comment]
> Possiblyxcb:marshall should eval it as well and pad the resulting message to be at least that size.

I don't think that's necessary. We'd have to set an optional field (e.g., switch) we don't understand (which doesn't make sense). If we're writing the datastructure, we'll know the length.

From the spec:

>   This makes it possible to handle structures with conditional fields (see the <switch> element) where the future revisions of protocols may introduce new variants and old code must still properly ignore them.

---

> Possibly xcb:unmarshall should consume additional bytes if less than ~size.

That should be what's happening here (by returning the size, same as for unions).

**Comment by Stebalien** [comment]
Ok, I've committed the float tests. We still need to wire everything up to makefiles and CI, but I'm going to do that in a separate PR after this one lands.

Everything should be good for a final round of reviews.

File: el_client.el [hunk 2]
```diff
@@ -63,6 +63,8 @@
 
 (defvar xelb-request-fields nil "Fields in the current request.")
 
+(defconst xelb-xproto-namespace "xproto:" "The namespace of the core protocol.")
```
Comment (minad) [inline el_client.el:66]: The usage of constants does not seem coherent in el_client.el. I've seen there is a variable xelb-prefix, but also many "xcb:" literal strings. But it seems to be the right thing to extract such constants like you have done here.

File: xcb-types.el [hunk 3]
```diff
@@ -456,6 +462,66 @@ Consider let-bind it rather than change its global value."))
...
+         (signmask (if isneg #x8000000000000000 0)))
+    (+ (cond ((zerop frac) 0)                                          ; 0
+             ((isnan frac) #xff0000000000001)                          ; NaN
+             ((or (>= exp 2047) (= frac 1e+INF)) #x7ff0000000000000)   ; Inf
+             ((<= exp 0) (ash (round (ldexp frac 52)) exp))            ; Subnormal
+             (t (+ (ash exp 52) (logand #xfffffffffffff
+                                        (round (ldexp frac 53))))))    ; Normal
+       signmask)))
+
+(defsubst xcb:-f32-to-binary32 (value)
```
Comment (minad) [inline xcb-types.el:481]: Just bikeshedding, but is binary32 commonly used as name for unpacked f32? We could also use unpack-f32 and pack-f32 to convert back and forth.

File: xcb-xprint.el [hunk 4]
```diff
@@ -198,11 +200,13 @@
 		 (xcb:-fieldref 'len-data))
 	  :type xcb:-list)
    (data :initarg :data :type xcb:-ignore)
+   (pad~0 :initform 4 :type xcb:-pad-align)
```
Comment (minad) [inline xcb-xprint.el:203]: Your script preserved almost all of the code which is good and expected when we are regenerating the files based on a new xcb protocol version. But I wonder about these paddings introduced here. Why haven't they been present before? Did the layouts change?

File: xelb-test.el [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    (#x0010000000000000 . 2.2250738585072014e-308)
+    (#x7fefffffffffffff . 1.7976931348623157e308)
+    (#x0000000000000000 . +0.0)
+    (#x8000000000000000 . -0.0)
+    (#x7ff0000000000000 . +1e+INF)
+    (#xfff0000000000000 . -1e+INF)
+    (#x3fd5555555555555 . ,(/ 1.0 3.0))
+    (#x400921fb54442d18 . ,float-pi)))
+
+(defmacro xelb-test--binary-n (n)
```
Comment (minad) [inline xelb-test.el:71]: Here I think I'd prefer a function instead of a macro. It could be passed three arguments (cases and the pack/unpack functions)-

**Comment by minad** [comment]
@Stebalien Thanks, this looks all good to me. I added a few comments.

File: xelb-test.el (hunk 5 above)
Comment (Stebalien) [reply xelb-test.el]: Fair enough. Fixed.

File: xcb-xprint.el (hunk 4 above)
Comment (minad) [reply xcb-xprint.el:203]: Thanks, found some of the relevant commits, e.g., https://cgit.freedesktop.org/xcb/proto/commit/src/xprint.xml?id=8a7096ce7375b1840aa350a27a092dd4a22f4c6e.

File: xcb-types.el (hunk 3 above)
Comment (Stebalien) [reply xcb-types.el:481]: 1. Naming: that's what IEEE 754 calls them call them.
2. Defsubst: There isn't a single defun in the entire file. IMO, we should be using more defuns, but I'd like to stick to the convention for now.

Comment (minad) [reply xcb-types.el:481]: Thanks, all good!

**Comment by Stebalien** [comment]
> You are right, the one constructing the message should make sure the right size is calculated (in xcb:xinput:DeviceClass one should adjust the len field as needed). The only case I think this could fail would be if one of the cases in the switch could not be expressed by the <length> expression. I would find it surprising if spec writers would do that... The xcb:marshall method of xcb:-union does extend the vector to fill the length, but there it's obviously needed. I defer this to your judgement.

I think the correct answer here is to assert that the size is correct, not to zero fill. Unions are "fixed size", so we need to pad to the correct size.

I've pushed a commit that does that (and adds some more paranoid checks to other sizes). If I'm wrong here, we should notice pretty quickly.

**Comment by Stebalien** [comment]
I think so, I haven't had any issues either.

**Pull request merged**